### ISSUE TITLE
Bug: Large number of html overlays causes openseadragon to crash

### ISSUE THREAD

**Issue by abarkellumich** [openseadragon/openseadragon#2008]
Hello, I am attempting to create a grid overlay with labels on various images, some quite large, with a fixed grid-spacing. I'm noticing that there seems to be a hard limit to the number of these html elements that I am able to load on a page, around 775 or so. The specific error when going over that number is:

Uncaught TypeError: this.size is undefined
    _getSizeInPixels openseadragon.js:17360
    _getOverlayPositionAndSize openseadragon.js:17332
    drawHTML openseadragon.js:17289
    _drawOverlays openseadragon.js:9230
    l openseadragon.js:7635
    success openseadragon.js:7701
    a openseadragon.js:8531
    addTiledImage openseadragon.js:8543
    success openseadragon.js:9356
    TileSource openseadragon.js:11692
    getHandler openseadragon.js:3073
    raiseEvent openseadragon.js:3095
    r openseadragon.js:12024
    success openseadragon.js:12046
    onreadystatechange openseadragon.js:2245
    
   
 I had also noticed some serious lag when navigating the image when I had a high number of overlays enabled, markedly more so on Chrome than FireFox. If anyone has some insights as to what is causing this, and or workarounds, it would be highly appreciated. I'm also more than happy to supply any additional information that you would find useful in diagnosing this, just let me know what you need.

**Comment by iangilman** [comment]
That's a lot of overlays! I'm not surprised you're getting performance issues.  

It looks like you're using the built-in overlays. I imagine you'll get better performance if you use one of the other overlay plugins, like maybe the SVG one, or possibly even one of the HTML overlay plugins. Another option would be to go with the canvas overlay plugin and draw different levels of detail at different zooms. 

As for the specific error you're getting, we certainly don't want it throwing exceptions. Do you have a reliable repro for that? Looking at the code I'm not sure that it would be based on having a large number of overlays, but I'm not sure what would cause it. Would you be up for digging into it further?

**Comment by abarkellumich** [comment]
> It looks like you're using the built-in overlays. I imagine you'll get better performance if you use one of the other overlay plugins, like maybe the SVG one, or possibly even one of the HTML overlay plugins. Another option would be to go with the canvas overlay plugin and draw different levels of detail at different zooms.

Correct, I'm using the built-in overlays. I was originally thinking along the same lines, we are using react as our front-end and I was having a fair bit of trouble getting those plugins to function properly. Maybe I'll take another look into those as well.


> As for the specific error you're getting, we certainly don't want it throwing exceptions. Do you have a reliable repro for that? Looking at the code I'm not sure that it would be based on having a large number of overlays, but I'm not sure what would cause it. Would you be up for digging into it further?

Sure, feel free to check this file out[0]. It's a bit messy as we've been doing a fair bit of exploratory work attempting to resolve this issue as well as get everything working inside react. The nuts and bolts of it are in the `initSeaDragon()` and `getGridOverlay()` functions.

Under `getGridOverlay()` There are two sets of for-loops, the first is to draw grid lines through supplying the x and y positions, line width, and line heights. The second set of loops generates specific labeling as for the html tags.


[0] https://github.com/KPMP/delphinus-web/blob/develop/src/components/Slides/SlideViewer.js

**Comment by iangilman** [comment]
Hmm... you should be able to get the plugins to work fine in React. I think it's worth trying again, from a performance standpoint. 

Thanks for link to the code. It looks reasonable... the one place I could imagine you might be getting that error is if this line: 

https://github.com/KPMP/delphinus-web/blob/66f61491474931c699fc636b8e3d179cd62ff2d9/src/components/Slides/SlideViewer.js#L154

...sometimes refers to an ID of an element that doesn't exist.
